### PR TITLE
controller: Clean up the Longhorn node when the kube node is gone

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#1232

### CODE REVIEW

**Comment by shuo-wu** [comment]
This PR is pretty straightforward. Not sure if I miss anything, and if we still need a LEP.

File: controller/kubernetes_node_controller.go [hunk 1]
```diff
@@ -198,11 +198,17 @@ func (knc *KubernetesNodeController) syncKubernetesNode(key string) (err error)
...
-			logrus.Errorf("Kubernetes node %v has been deleted", key)
-			return nil
+		if !datastore.ErrorIsNotFound(err) {
+			return err
 		}
-		return err
+		logrus.Errorf("Kubernetes node %v has been deleted", key)
+	}
+
+	if kubeNode == nil || kubeNode.DeletionTimestamp != nil {
```
Comment (yasker) [inline controller/kubernetes_node_controller.go:207]: Should we count `DeletionTimestamp != nil`? What if the manager is still running on the node?

Also, do we need to do the replica cleanup after the node deletion?

Comment (shuo-wu) [reply controller/kubernetes_node_controller.go:207]: OK. Maybe I cannot count `DeletionTimestamp != nil`. And there are some issues/corner cases for the replicas on the node. I will need to re-check those cases.

**Comment by shuo-wu** [comment]
The following cases should be tested:
1. When the node is gone,
    1. -[ ] The volume attached to the node should become and keep the state `Faulted`. And there is no need to detach the volume, which is similar to the node-down case.
        1. -[ ] The only replica is on the removed node, the volume won't be salvaged.
    2. -[ ] The volume is not attached to the removed node. The replica on the removed node should become ERROR.
2. After the node is removed,
    1. -[ ] The replicas on the removed node cannot be salvaged.
    2. -[ ] All volumes can be detached and reattached manually.
3. -[ ] If a node with the same node is added to the cluster, all volumes can be recovered automatically unless the only replica becomes ERROR

**Review by yasker — APPROVED** [review]

**Comment by yasker** [comment]
LGTM. Let's wait for the test result.

**Pull request merged**